Save aliases in INSTALL_RECEIPT

penman opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This is part of #567 (but not all of it ⏤ I wanted to PR this bit so that I could get feedback since I haven't worked with this codebase before).

This PR adds an `install_name` key to INSTALL_RECEIPT. Usually this will just be set to the name of the formula, but if it was installed as an alias, the `install_name` will be set to the alias instead.

Implementation-wise, I wasn't sure about how best to set the `install_name` on a `Formula` instance (which seemed like the only appropriate place to store it). I thought the best way would probably be a keyword argument, but it could also be a standard optional argument, or an `attr_accessor`. I'm open to suggestions if there's a better way.

penman commented:
Looks like that CI failure isn't anything to do with my changes? I ran `brew test-bot` just to make sure and everything passed.

MikeMcQuaid commented:
> Looks like that CI failure isn't anything to do with my changes? I ran brew test-bot locally just to make sure and everything passed.

Yeh, the Travis build has been flaky this weekend. The `default` is the main one we care about.

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -143,9 +148,10 @@ class Formula
 
   # @private
-  def initialize(name, path, spec)
+  def initialize(name, path, spec, install_name: name)
```
We're still mostly using hash rockets until we mass change them at some later point. Were still on Ruby 1.8 until recently!

penman replied:
Doesn't look like it's possible to use hash rockets to declare a keyword argument.

penman replied:
(I'll change the calls to `new` to use hash rockets, though.)

MikeMcQuaid replied:
Ah, my mistake, missed this was a kwarg. Fine as-is then 👍

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -502,6 +508,21 @@ def installed_kegs
+  # tapped, so we don't have to worry about doing extra HTTP requests, or other
+  # expensive operations, when looking them up again.
+  def install_ref
+    install_name == name ? path : install_name
```
This will probably need to be a fully-qualified path.

MikeMcQuaid replied:
or alternatively you could return `name` or `install_name` rather than `path` and build the fully-qualified one where it's needed?

penman replied:
Didn't occur to me that I could return the path to the alias, but it's obvious in hindsight!

MikeMcQuaid commented on Library/Homebrew/formulary.rb:
```diff
@@ -68,15 +68,18 @@ class FormulaLoader
     # Gets the formula instance.
     def get_formula(spec)
-      klass.new(name, path, spec)
+      klass.new(name, path, spec, install_name: install_name)
```
Similarly here wrt hash rockets.

MikeMcQuaid commented on Library/Homebrew/formulary.rb:
```diff
@@ -118,6 +121,7 @@ def initialize(alias_path)
       path = alias_path.resolved_path
       name = path.basename(".rb").to_s
       super name, path
+      @install_name = alias_path.basename.to_s
```
I was wondering if things would 💥 if we used the `@install_name` as the actual value for `name` instead? That means we could avoid having to add new fields. Thoughts?

penman replied:
I was afraid to try. :P

I imagine `Formula#name` is used pretty widely across Homebrew. This would, for example, change the install prefix for a formula based on the alias used to install it, which doesn't seem like a great idea to me.

MikeMcQuaid replied:
A good argument 👍

MikeMcQuaid replied:
Could maybe be something like `unresolved_name` or something?

MikeMcQuaid replied:
(naming is hard)

penman replied:
`unresolved_name` would imply to me that it could be a URL, for a file path, which currently it'll never be…

MikeMcQuaid replied:
`alias_name`?

MikeMcQuaid replied:
or, if we went with the full path for above, `alias_path`

penman replied:
How about we go for `alias_path`, and make it return `nil` rather than the formula path if an alias wasn't used?

Then we could use `formula.alias_path || formula.path` when passing a reference to it, and the meaning would be clearer all round.

MikeMcQuaid replied:
Sounds good 👍.

MikeMcQuaid commented:
Some comments on this but liking where this is going so far and nice work creating this PR!

MikeMcQuaid commented on Library/Homebrew/formula_installer.rb:
```diff
@@ -589,7 +589,7 @@ def build
       --
       #{HOMEBREW_LIBRARY_PATH}/build.rb
-      #{formula.path}
+      #{formula.alias_path || formula.path}
```
Think this logic may be worth pulling into a formula method that just does this `||`. `alias_path_or_path` or something with a better name?

penman replied:
I thought about this. My reasoning for leaving it as-is is that having a method called `alias_path_or_path` wouldn't really be reducing repetition at all. It just occurred to me that `specified_path` would work as a name, though, so I'll move it into a method.

MikeMcQuaid commented on Library/Homebrew/tab.rb:
```diff
@@ -292,6 +298,7 @@ def to_json
       "time" => time,
       "source_modified_time" => source_modified_time.to_i,
       "HEAD" => self.HEAD,
+      "alias_path" => alias_path && alias_path.to_s,
```
Probably worth doing `(alias_path.to_s if alias_path)` just to match style below.

MikeMcQuaid commented:
Nice work today ✨

MikeMcQuaid commented on Library/Homebrew/tab.rb:
```diff
@@ -161,6 +166,7 @@ def self.empty
       "time" => nil,
       "source_modified_time" => 0,
       "HEAD" => nil,
+      "alias_path" => nil,
```
I'd maybe put this in `source` below and perhaps consider overriding `path` there unless it would break existing functionality relying on this in the tab.

penman replied:
As far as I can tell nothing uses source.path, so this should be 👍

MikeMcQuaid commented on Library/Homebrew/tab.rb:
```diff
@@ -292,6 +298,7 @@ def to_json
       "time" => time,
       "source_modified_time" => source_modified_time.to_i,
       "HEAD" => self.HEAD,
+      "alias_path" => (alias_path.to_s if alias_path),
```
Same here.

MikeMcQuaid commented:
A couple more tiny nits but I ❤️ how this is looking and when you're testing is 🆗 then I'd be happy to merge this.

penman commented:
I've updated the PR based on your feedback, @MikeMcQuaid.

It also looked to me like putting data into a Tab isn't actually tested anywhere, so I added `test_create` and `test_for_formula` to make sure the `source.path` is set correctly in both places. I could expand this to test all of the Tab attributes if you're interested?

MikeMcQuaid commented:
@penman This is looking great 👏. How's your testing going? When you consider this ready to 🚢 give me what you've been using for testing and I'll try my own too and, if it all looks good, we 🚢 this thing!

> I could expand this to test all of the Tab attributes if you're interested?

Always interested in more tests! Feel free to make that a future PR, though, to avoid blocking this one.

penman commented:
@MikeMcQuaid I'm skeptical of that test failure. I'm not sure how this change could break <kbd>cask cleanup</kbd>, and I can't reproduce it locally. Is it something you've seen before?

MikeMcQuaid commented:
@penman Yeh, ignore it, it's flaky.

jawshooah commented:
@reitermarkus Looks like [this spec](https://github.com/Homebrew/brew/blob/911edb0ed77859445c3a48f6abe5b9be5807ebf7/Library/Homebrew/cask/spec/cask/cli/cleanup_spec.rb#L56-L74) isn't entirely consistent. Mind taking a look?

penman commented:
@MikeMcQuaid In that case, I think this is ready.

Mostly, what I've been doing to test is installing <kbd>the_silver_searcher</kbd>, checking the INSTALL_RECEIPT has `source.path` set to the path to the formula, then uninstalling and installing it with <kbd>ag</kbd> (an alias for <kbd>the_silver_searcher</kbd>) and checking `source.path` is set to the path to the alias. I've tried a similar thing with a few other formulae with aliases, and it seems to be working. I've also installed some formulae with no aliases and that's been working too.

Nothing has come up in my general usage of Homebrew, either.

I'm curious to see what you test, though, because I'm sure there are paths I haven't thought of where something might have gone wrong, or maybe there are parts of Homebrew that are generally tricky when things change that I'm not aware of.

MikeMcQuaid commented:
Looks good to me in local testing. Nice work!

MikeMcQuaid commented:
Rerunning CI then will merge.

MikeMcQuaid commented:
@jawshooah @reitermarkus Going to need to remove that spec soon; it's making it pretty hard to get consistently green Homebrew builds.
